src/shm/pool/multi.rs

```diff
@@ -0,0 +1,159 @@
+        }
+    }
+    /// Returns the buffer associated with the given surface.
+    /// If the buffer isn't free, it returns None.
+    pub fn create_buffer<D, U>(
+        &mut self,
+        width: i32,
+        stride: i32,
+        height: i32,
+        identifier: &I,
+        format: wl_shm::Format,
+        udata: U,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+    ) -> Option<(usize, wl_buffer::WlBuffer, &mut [u8])>
+    where
+        D: Dispatch<wl_buffer::WlBuffer, UserData = U> + 'static,
+        U: Send + Sync + 'static,
+    {
+        let mut found_identifier = false;
+        let mut offset = 0;
+        let size = stride * height;
+        let mut index = None;
+        for (i, buffer) in self.buffer_list.iter_mut().enumerate() {
+            if buffer.identifier.eq(identifier) {
+                found_identifier = true;
+                if buffer.free.load(Ordering::Relaxed) {
+                    buffer.size = buffer.size.max(size as usize);
+                    index = Some(i);
+                }
+            } else if offset > buffer.offset {
+                if buffer.free.load(Ordering::Relaxed) {
+                    buffer.offset = offset;
+                } else {
+                    index = None;
+                }
+            } else if found_identifier {
+                break;
+            }
+            offset += buffer.size;
+        }
+
+        if found_identifier {
+            offset = self.buffer_list[index?].offset;
+        } else if let Some(b) = self.buffer_list.last() {
+            offset += b.size;
+        }
+
+        if offset + size as usize > self.inner.len
+        && self.resize(offset + 2 * size as usize, conn).is_err() {
+            return None
+        }
+
+        let buffer = self.inner.create_buffer(
+            offset as i32,
+            width,
+            height,
+            stride,
+            format, udata, conn,
+            qh
+        ).ok()?;
+
+        if found_identifier {
+            self.buffer_list[index?].buffer = buffer.clone();
+        } else if index.is_none() {
+            index = Some(self.buffer_list.len());
+            let buffer = Buffer {
+                offset,
+                free: AtomicBool::new(true),
+                buffer: buffer.clone(),
+                size: size as usize,
+                identifier: identifier.clone()
+            };
+            self.buffer_list.push(buffer);
+        }
+
+        let slice = &mut self.inner.mmap()[offset..][..size as usize];
+
+        self.buffer_list[index?].free.swap(false, Ordering::Relaxed);
+
+        Some((offset, buffer, slice))
+    }
+}
+
+impl<I: PartialEq + Clone> DelegateDispatchBase<wl_buffer::WlBuffer> for MultiPool<I> {
+    type UserData = ();
+}
+
+impl<D, I: PartialEq + Clone> DelegateDispatch<wl_buffer::WlBuffer, D> for MultiPool<I>
+where
+    D: Dispatch<wl_buffer::WlBuffer, UserData = Self::UserData>,
+    D: AsMut<MultiPool<I>>
```

Comment:
This introduces an undue restriction on a number of pools you can have. For example, a pool per window is probably fine. I imagine the solution here is a sort of user data on the buffer that signals the pool it was freed.

Reply:
I'm not sure about the meaning of undue restrictions. Is it too much or too little restrictions? I added this so n buffering would be possible with multiple surfaces on the same mempool depending on the identifier. If it's not necessary I'll revert back.

Reply:
AsMut contextually cannot select the right pool if you have more than one.

Reply:
I see. Maybe a trait like AsMutFrom<T, U> with U being the WlBuffer would work? This way the user could find the pool which holds the buffer.